File: consumer-node/src/pluginLoader.js
const fs = require("fs");
const path = require("path");
const BasePlugin = require("./basePlugin");
const pino = require("pino");

const log = pino({ name: "plugin-loader" });

const PLUGINS_DIR = path.join(__dirname, "plugins");

/**
 * Scan the plugins/ directory, require every .js file, and collect
 * instances of BasePlugin subclasses.  Adding a new plugin is as simple
 * as dropping a new file into plugins/ — no changes needed here.
 */
function discoverPlugins() {
  const instances = [];

  const files = fs.readdirSync(PLUGINS_DIR).filter((f) => f.endsWith(".js")).sort();

  for (const file of files) {
    const fullPath = path.join(PLUGINS_DIR, file);
    try {
      const PluginClass = require(fullPath);

      if (typeof PluginClass === "function" && PluginClass.prototype instanceof BasePlugin) {
        const instance = new PluginClass();
        instances.push(instance);
        log.info({ plugin: instance.name }, "Loaded plugin");
      }
    } catch (err) {
      log.error({ file, err }, "Failed to load plugin module");
    }
  }

  log.info({ count: instances.length }, "Total plugins discovered");
  return instances;
}

module.exports = { discoverPlugins };

File: consumer-python/plugin_loader.py
import os
import importlib
import inspect
import logging

from base_plugin import BasePlugin

log = logging.getLogger("plugin_loader")

PLUGINS_DIR = os.path.join(os.path.dirname(__file__), "plugins")


def discover_plugins() -> list[BasePlugin]:
    """
    Scan the plugins/ directory, import every .py module, and instantiate
    all concrete BasePlugin subclasses found.  No changes to this function
    are needed when a new plugin is added — just drop a file into plugins/.
    """
    instances = []

    for filename in sorted(os.listdir(PLUGINS_DIR)):
        if filename.startswith("_") or not filename.endswith(".py"):
            continue

        module_name = f"plugins.{filename[:-3]}"
        try:
            mod = importlib.import_module(module_name)
        except Exception:
            log.exception("Failed to import plugin module %s", module_name)
            continue

        for attr_name in dir(mod):
            attr = getattr(mod, attr_name)
            if (
                inspect.isclass(attr)
                and issubclass(attr, BasePlugin)
                and attr is not BasePlugin
                and not inspect.isabstract(attr)
            ):
                try:
                    instance = attr()
                    instances.append(instance)
                    log.info("Loaded plugin: %s", instance.name)
                except Exception:
                    log.exception("Failed to instantiate plugin %s", attr_name)

    log.info("Total plugins discovered: %d", len(instances))
    return instances
